### PR TITLE
Shortcode parser fixes

### PR DESCRIPTION
**Pull request by luchaos**
- Fix empty shortcode tags causing errors -> #1009
- Only fix url protocol for missing scheme and retroachievment URLs -> #957

### CODE REVIEW

File: src/Shortcode.php [hunk 1]
```diff
@@ -91,13 +91,18 @@ private function renderLink(ShortcodeInterface $shortcode): string
...
+            return '';
+        }
+
         $scheme = parse_url($href, PHP_URL_SCHEME);
+        $host = parse_url($href, PHP_URL_HOST);
 
         if (empty($scheme)) {
             $href = 'https://' . ltrim($href, '/');
-        } elseif ($scheme === 'http') {
+        } elseif ($scheme === 'http' && str_ends_with($host, 'retroachievements.org')) {
```
Comment (luchaos) [inline src/Shortcode.php:105]: Only our own links will be https-only.

File: src/Shortcode.php [hunk 2]
```diff
@@ -177,10 +182,37 @@ private function embedTicket($id): string
...
+    private function autolinkRetroachievementsUrls(string $text): string
+    {
+        // see https://stackoverflow.com/a/2271552/580651:
+        // [...] it's probably safe to assume a semicolon at the end of a URL is meant as sentence punctuation.
+        // The same goes for other sentence-punctuation characters like periods, question marks, quotes, etc..
+        // lookahead: (?<![!,.?;:"\'()-])
+        return (string) preg_replace(
+            '~
+                (?:https?://)?      # Optional scheme. Either http or https.
+                (?:www.)?           # Optional subdomain.
```
Comment (luchaos) [inline src/Shortcode.php:201]: www subdomains for retroachievements.org will be dropped

File: src/Shortcode.php [hunk 3]
```diff
@@ -177,10 +182,37 @@ private function embedTicket($id): string
...
+                retroachievements\.        # Host.
+                ([\w!#$%&\'()*+,./:;=?@\[\]-]+
+                (?<![!,.?;:"\'()]))
+                (?!                 # Assert URL is not pre-linked.
+                  [^<>]*>           # Either inside a start tag,
+                  | [^<>]*</a>      # End recognized pre-linked alts.
+                )                   # End negative lookahead assertion.
+            ~ix',
+            '<a href="https://retroachievements.$1">https://retroachievements.$1</a>',
```
Comment (Jamiras) [inline src/Shortcode.php:210]: My regex-fu is not very strong, but wouldn't `$1` be the `https` or `www` capture?

Comment (luchaos) [reply src/Shortcode.php:210]: Same here. Yet, I learned a while ago that `(?` are non-capturing groups; meaning they will be treated as an entity but not exposed as a match. I trusted regex101, stackoverflow and the tests in this case tbh.
The reason for not including `org` in the domain stems from pretty much the same vein of not having a damn clue what regex wants from me. The sentence-punctuation characters clause stopped working as soon as i wanted to include the tld in there for some reason. Chose the easy way out. But it works 🤷

**Review by Jamiras — APPROVED** [review]

**Pull request merged**